File: NHSOnline.Backend.Worker.IntegrationTests/Worker/NhsoHttpException.cs
using System;
using System.Net;
using System.Net.Http;
using System.Text;

namespace NHSOnline.Backend.Worker.IntegrationTests.Worker
{
    public class NhsoHttpException : Exception
    {
        public NhsoHttpException(HttpRequestMessage request, HttpResponseMessage response)
        {
            StatusCode = response.StatusCode;
            Body = response.ToString();
            RequestUri = request.RequestUri;
            Method = request.Method;
        }

        public HttpMethod Method { get; set; }
        public string Body { get; }
        public Uri RequestUri { get; }
        public HttpStatusCode StatusCode { get; }

        public override string ToString()
        {
            var builder = new StringBuilder();
            builder.AppendLine("Stubs threw an HTTP exception:");
            builder.AppendLine($"  Status code: {StatusCode}");
            builder.AppendLine($"  Request URI: {RequestUri}");
            builder.AppendLine($"  Method:      {Method}");
            builder.AppendLine("  Body:");
            builder.AppendLine(Body);
            return builder.ToString();
        }
    }
}

File: src/main/kotlin/worker/NhsoHttpException.kt
package worker

import org.apache.http.HttpRequest
import org.apache.http.HttpResponse

class NhsoHttpException(request: HttpRequest, response: HttpResponse, var method: String? = null) : RuntimeException() {

    val Body: String = response.toString()
    val RequestUri: String = request.requestLine.uri
    var StatusCode: Int = response.statusLine.statusCode

    init {
        method = request.requestLine.method
    }

    override fun toString(): String {
        val builder = StringBuilder()
        builder.append("Stubs threw an HTTP exception:" + "\r\n")
        builder.append(String.format("  Status code: %1\$s", StatusCode) + "\r\n")
        builder.append(String.format("  Request URI: %1\$s", RequestUri) + "\r\n")
        builder.append(String.format("  Method:      %1\$s", method) + "\r\n")
        builder.append("  Body:" + "\r\n")
        builder.append(Body + "\r\n")
        return builder.toString()
    }
}